### PR TITLE
Fix #907: Add Enrollments to the unittest genesis block

### LINKED ISSUE #907: Add Enrollments to the hardcoded genesis blocks
Depends on #821

Our genesis blocks need to have Enrollments in themselves in order to be valid and usable.
Currently the lack of Enrollment means we have to come up with weird workaround for our tests.
Issues such as #822 ended up being slowed down by this.

### PR DESCRIPTION
**Pull request by Geod24**
See each commits for details.

### CODE REVIEW

File: source/agora/consensus/data/Block.d [hunk 1]
```diff
@@ -240,8 +240,6 @@ public struct Block
     }
     do
     {
-        () @trusted { merkle_tree.assumeSafeAppend(); }();
-
```
Comment (Geod24) [inline source/agora/consensus/data/Block.d:244]: Split in its own commit

File: source/agora/consensus/Genesis.d [hunk 2]
```diff
@@ -140,7 +151,7 @@ unittest
 unittest
 {
     import agora.common.Serializer;
-    testSymmetry(UnitTestGenesisTransaction);
+    testSymmetry(UnitTestGenesisBlock.txs[0]);
```
Comment (Geod24) [inline source/agora/consensus/Genesis.d:187]: Foreach directly

File: source/agora/consensus/Genesis.d [hunk 3]
```diff
@@ -122,13 +147,22 @@ unittest
...
-         `3dc369ef2a44b62ba113814a9d819a276ff61582874c9aee9c98efa2aa1f10d73`);
+private immutable Hash[] UnitTestGenesisMerkleTree = [
+    Hash(`0x6314ce9bc41a7f5b98309c3a3d824647d7613b714c4e3ddbc1c5e9ae46db297` ~
+         `15c83127ce259a3851363bff36af2e1e9a51dfa15c36a77c9f8eba6826ff975bc`),
+    Hash(`0x7a5bfeb96f9caefa377cb9a7ffe3ea3dd59ea84d4a1c66304ab8c307a4f4770` ~
+         `6fe0aec2a73ce2b186a9f45641620995f8c7e4c157cee7940872d96d9b2f0f95c`),
+    Hash(`0x788c159d62b565655d9f725786c38e6802038ee73d7a9d187b3be1c7de95aa0` ~
+         `ba856bf81bb556d7448488e71f4b89ce6eba319d0536798308112416413289254`),
+];
+
```
Comment (Geod24) [inline source/agora/consensus/Genesis.d]: Split / move

**Review by Geod24 — COMMENTED** [review]
Needs a re-order / split

File: source/agora/consensus/Genesis.d (hunk 2 above)
Comment (Geod24) [reply source/agora/consensus/Genesis.d:187]: Actually no, since we do the whole block anyway

**Comment by Geod24** [comment]
Rebased on `v0.x.x`, cleaned up, ready for review.

File: source/agora/consensus/Genesis.d [hunk 4]
```diff
@@ -72,10 +73,22 @@ public void setGenesisBlock (immutable Block* block)
...
+/*******************************************************************************
+
+    The genesis block as used by most unittests
+
+    Note that this is more of a 'test' block than a 'unittest' block,
+    and it's currently used in a few integration test, hence why it is not
+    `version (unittest)`.
+
+    It contains a total of 500M initial coins, of which 12M have been frozen
+    amount 6 nodes, and the rest is evenly split between 8 outputs (61M each).
```
Comment (AndrejMitrovic) [inline source/agora/consensus/Genesis.d]:
```suggestion
    among 6 nodes, and the rest is evenly split between 8 outputs (61M each).
```

Comment (AndrejMitrovic) [reply source/agora/consensus/Genesis.d]: or `for 6 nodes each`

**Review by AndrejMitrovic — APPROVED** [review]

File: source/agora/consensus/Genesis.d (hunk 4 above)
Comment (Geod24) [reply source/agora/consensus/Genesis.d]: Fixed

**Comment by AndrejMitrovic** [comment]
It was green, merging.

**Comment by AndrejMitrovic** [comment]
Actually this PR was incomplete. There were two checks missing (including a bug fix):

```diff
diff --git a/source/agora/consensus/Genesis.d b/source/agora/consensus/Genesis.d
index 5d5de3e5..fa30549b 100644
--- a/source/agora/consensus/Genesis.d
+++ b/source/agora/consensus/Genesis.d
@@ -57,7 +57,7 @@ public void setGenesisBlock (immutable Block* block)
 {
     assert(block !is null);

-    if (auto reason = isGenesisBlockInvalidReason(*gen_block))
+    if (auto reason = isGenesisBlockInvalidReason(*block))
         throw new Exception(reason);

     gen_block = block;
diff --git a/source/agora/consensus/validation/Block.d b/source/agora/consensus/validation/Block.d
index f6db88cd..939d2b6a 100644
--- a/source/agora/consensus/validation/Block.d
+++ b/source/agora/consensus/validation/Block.d
@@ -198,7 +198,7 @@ public string isGenesisBlockInvalidReason (const ref Block block) nothrow @safe

     // If there are no enrollments, return them here early.
     if (block.header.enrollments.length == 0)
-        return null;
+        return "GenesisBlock: There must be at least one enrollment";

     if (!isStrictlyMonotonic!"a.utxo_key < b.utxo_key"
         (block.header.enrollments))
```

Also the validation still fails. Because there are still no enrollments in the genesis block. You only added freeze transactions.

**Comment by Geod24** [comment]
🤦
